pwnlib/heap/glmalloc/process_informer.py

```diff
@@ -0,0 +1,51 @@
+import os.path
+import pwnlib.util.proc
+from pwnlib.heap.glmalloc.utils import \
+    u64, \
+    u32, \
+    get_libc_version_from_name, \
+    get_main_arena_addr
+
+
+class ProcessInformer:
+    """Helper class with information about process memory such as pointer_size,
+    and which allows perform operations over memory.
+
+    Attributes:
+        libc_version (tuple(int, int)): The glibc version in
+            format (major, minor)
+        pid (int): pid of the process
+        pointer_size (int): size in bytes of a pointer in the process
+        main_arena_address (int): address of the main arena malloc state
+
+    """
+
+    def __init__(self, pid, libc):
+        libc_name = os.path.basename(libc.path)
+        self.libc_version = get_libc_version_from_name(libc_name)
+        self.pid = pid
+
+        if "64" in libc.get_machine_arch():
+            self.pointer_size = 8
+            self.unpack_pointer = u64
+        else:
+            self.pointer_size = 4
+            self.unpack_pointer = u32
+
+        self.unpack_int = u32
+
+        self.main_arena_address = get_main_arena_addr(libc, self.pointer_size)
+
+    def read_memory(self, address, size):
```

Comment:
We should not be doing manual reading from /proc/xxx/mem for several reasons:

* It does not exist on some Unixes
* It may not be accessible at all (in a CTF environment)
* Corefile support is already rich and gives you all the read/write you could want

Reply:
Finally, I've created two interfaces to get the heap data, one for corefile (CorefileInformer)  and another for process (ProcessInformer). You can use the corefile interface for those cases you mentioned, and for testing too. 

However I have left the process interface also with some changes since if you are analyzing a live process, creating a corefile for each heap change is time expensive and very annoying since requires you to instantiate a HeapExplorer for each heap change, which is counter-intuitive, and you create corefile per process which may left your directory full of corefiles. 

Then you can use process.heap_explorer() for live analysis or process.corefile.heap_explorer() for analyze the corefile, as you wish.